File: ios/Sources/Protocol.swift
// Mirror of shared/protocol.ts — JSON messages discriminated by `type`.

import Foundation

struct PlayerState: Decodable {
    let id: String
    let name: String
    let pos: [Double]
    let yaw: Double
    let pitch: Double
    let anim: Int
    let hp: Int
    let ammo: Int
    let reserve: Int
    let kills: Int
    let boarded: Bool

    var isDead: Bool { anim & 16 != 0 }
}

struct BugState: Decodable {
    let id: Int
    let kind: Int
    let pos: [Double]
    let yaw: Double
    let hp: Int
}

struct SupplyState: Decodable {
    let id: Int
    let pos: [Double]
    let charges: Int
}

struct MissionState: Decodable {
    let phase: String
    let seed: UInt32
    let kills: Int
    let killTarget: Int
    let reinforceLeft: Int
    let phaseEndsAt: Double
}

enum ServerMsg {
    case welcome(selfId: String, room: String, host: String, mission: MissionState, players: [PlayerState])
    case errorMsg(reason: String)
    case joined(player: PlayerState)
    case left(id: String, host: String)
    case snapshot(t: Double, players: [PlayerState], bugs: [BugState], supplies: [SupplyState], mission: MissionState)
    case fired(id: String, origin: [Double], dir: [Double], hit: [Double]?, hitKind: String)
    case bugDeath(id: Int, pos: [Double], kind: Int)
    case hit(id: String, hp: Int)
    case down(id: String)
    case strat(id: String, kind: String, target: [Double], at: Double)
    case boom(kind: String, pos: [Double])
    case reinforced(ids: [String], pos: [Double])
    case phase(mission: MissionState)
    case boarded(id: String)
    case pong(t: Double, now: Double)
    case unknown(type: String)
}

extension ServerMsg: Decodable {
    private enum CodingKeys: String, CodingKey {
        case type, `self`, room, host, mission, players, reason, player, id
        case t, bugs, supplies, origin, dir, hit, hitKind, pos, kind, hp
        case target, at, ids, now
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "welcome":
            self = .welcome(
                selfId: try c.decode(String.self, forKey: .self),
                room: try c.decode(String.self, forKey: .room),
                host: try c.decode(String.self, forKey: .host),
                mission: try c.decode(MissionState.self, forKey: .mission),
                players: try c.decode([PlayerState].self, forKey: .players)
            )
        case "error":
            self = .errorMsg(reason: try c.decode(String.self, forKey: .reason))
        case "joined":
            self = .joined(player: try c.decode(PlayerState.self, forKey: .player))
        case "left":
            self = .left(
                id: try c.decode(String.self, forKey: .id),
                host: try c.decode(String.self, forKey: .host)
            )
        case "snapshot":
            self = .snapshot(
                t: try c.decode(Double.self, forKey: .t),
                players: try c.decode([PlayerState].self, forKey: .players),
                bugs: try c.decode([BugState].self, forKey: .bugs),
                supplies: try c.decode([SupplyState].self, forKey: .supplies),
                mission: try c.decode(MissionState.self, forKey: .mission)
            )
        case "fired":
            self = .fired(
                id: try c.decode(String.self, forKey: .id),
                origin: try c.decode([Double].self, forKey: .origin),
                dir: try c.decode([Double].self, forKey: .dir),
                hit: try c.decodeIfPresent([Double].self, forKey: .hit),
                hitKind: try c.decode(String.self, forKey: .hitKind)
            )
        case "bugDeath":
            self = .bugDeath(
                id: try c.decode(Int.self, forKey: .id),
                pos: try c.decode([Double].self, forKey: .pos),
                kind: try c.decode(Int.self, forKey: .kind)
            )
        case "hit":
            self = .hit(
                id: try c.decode(String.self, forKey: .id),
                hp: try c.decode(Int.self, forKey: .hp)
            )
        case "down":
            self = .down(id: try c.decode(String.self, forKey: .id))
        case "strat":
            self = .strat(
                id: try c.decode(String.self, forKey: .id),
                kind: try c.decode(String.self, forKey: .kind),
                target: try c.decode([Double].self, forKey: .target),
                at: try c.decode(Double.self, forKey: .at)
            )
        case "boom":
            self = .boom(
                kind: try c.decode(String.self, forKey: .kind),
                pos: try c.decode([Double].self, forKey: .pos)
            )
        case "reinforced":
            self = .reinforced(
                ids: try c.decode([String].self, forKey: .ids),
                pos: try c.decode([Double].self, forKey: .pos)
            )
        case "phase":
            self = .phase(mission: try c.decode(MissionState.self, forKey: .mission))
        case "boarded":
            self = .boarded(id: try c.decode(String.self, forKey: .id))
        case "pong":
            self = .pong(
                t: try c.decode(Double.self, forKey: .t),
                now: try c.decode(Double.self, forKey: .now)
            )
        default:
            self = .unknown(type: type)
        }
    }
}

// Client → server messages are tiny; build them as dictionaries.
enum ClientMsg {
    static func join(name: String, room: String?) -> [String: Any] {
        var m: [String: Any] = ["type": "join", "v": K.protocolVersion, "name": name]
        if let room, !room.isEmpty { m["room"] = room }
        return m
    }
    static func start() -> [String: Any] { ["type": "start"] }
    static func state(pos: [Double], yaw: Double, pitch: Double, anim: Int) -> [String: Any] {
        ["type": "state", "pos": pos, "yaw": yaw, "pitch": pitch, "anim": anim]
    }
    static func fire(origin: [Double], dir: [Double]) -> [String: Any] {
        ["type": "fire", "origin": origin, "dir": dir]
    }
    static func reload() -> [String: Any] { ["type": "reload"] }
    static func stratagem(kind: String, target: [Double]) -> [String: Any] {
        ["type": "stratagem", "kind": kind, "target": target]
    }
    static func interact() -> [String: Any] { ["type": "interact"] }
    static func ping(t: Double) -> [String: Any] { ["type": "ping", "t": t] }
}

File: shared/protocol.ts
// Wire protocol: JSON messages discriminated by `type`.
// This file is the single source of truth; the Swift client mirrors it in
// ios/Sources/Protocol.swift. v1 favors debuggability over bandwidth — the
// upgrade path is MessagePack/binary snapshots behind the same shapes.

export type Vec3 = [number, number, number];

export const ANIM = {
  MOVING: 1,
  SPRINT: 2,
  FIRING: 4,
  RELOADING: 8,
  DEAD: 16,
  STRAT: 32, // punching in a stratagem code
  DIVING: 64, // mid dive/dodge roll
} as const;

export interface PlayerState {
  id: string;
  name: string;
  pos: Vec3;
  yaw: number;
  pitch: number;
  anim: number;
  hp: number;
  ammo: number;
  reserve: number;
  kills: number;
  boarded: boolean;
  weapon?: string; // equipped weapon id
  level?: number; // account level (for squad display)
}

// A player's persistent account snapshot, sent on join and after each mission.
export interface ProfileState {
  level: number;
  xp: number; // cumulative lifetime XP
  into: number; // XP earned into the current level
  span: number; // XP the current level spans
  unlocked: string[]; // weapon ids available at this level
  kills: number; // lifetime kills
  missions: number; // lifetime missions completed
}

export interface BugState {
  id: number;
  kind: number; // index into BUG_KINDS
  pos: Vec3;
  yaw: number;
  hp: number;
  flags?: number; // BUGFLAG bits (windup/charging), omitted when 0
}

export const BUGFLAG = {
  WINDUP: 1, // charger telegraphing a lunge
  CHARGING: 2, // charger mid-lunge
} as const;

export interface ProjectileState {
  id: number;
  kind: number; // 0 acid glob (spitter), 1 bile (titan)
  pos: Vec3;
  vel: Vec3; // lets clients extrapolate between snapshots
}

export interface SupplyState {
  id: number;
  pos: Vec3;
  charges: number;
}

export interface SentryState {
  id: number;
  pos: Vec3;
  yaw: number;
  hp: number;
}

export interface FireZoneState {
  id: number;
  pos: Vec3;
  radius: number;
}

export type MissionPhase =
  | 'LOBBY'
  | 'DROP'
  | 'KILL'
  | 'EXTRACT'
  | 'DEFEND'
  | 'BOARD'
  | 'COMPLETE'
  | 'FAILED';

export type MissionObjective = 'ERADICATE' | 'NESTS';

export interface MissionState {
  phase: MissionPhase;
  seed: number;
  objective: MissionObjective;
  kills: number;
  killTarget: number;
  nestsLeft: number; // live bug nests (NESTS objective)
  nestsTotal: number;
  reinforceLeft: number;
  // server clock (ms) when the current phase auto-advances; 0 if not timed
  phaseEndsAt: number;
}

export interface NestState {
  i: number; // index into the layout's nests
  pos: Vec3;
  hp: number;
  hpMax: number;
}

export type ClientMsg =
  | { type: 'join'; v: number; name: string; room?: string; pid?: string } // pid = persistent account id
  | { type: 'start' }
  | { type: 'state'; pos: Vec3; yaw: number; pitch: number; anim: number }
  | { type: 'fire'; origin: Vec3; dir: Vec3 }
  | { type: 'dive'; dir: Vec3 }
  | { type: 'loadout'; weapon: string; stratagems: string[] } // chosen in the lobby
  | { type: 'reload' }
  | { type: 'stratagem'; kind: string; target: Vec3 }
  | { type: 'interact' }
  | { type: 'ping'; t: number };

export type HitKind = 'bug' | 'player' | 'rock' | 'nest' | 'none';

export type ServerMsg =
  | {
      type: 'welcome';
      self: string;
      room: string;
      host: string;
      mission: MissionState;
      players: PlayerState[];
      profile: ProfileState; // the joining player's account
    }
  | { type: 'error'; reason: string }
  | { type: 'joined'; player: PlayerState }
  | { type: 'left'; id: string; host: string }
  | {
      type: 'snapshot';
      t: number; // server clock ms
      players: PlayerState[];
      bugs: BugState[];
      supplies: SupplyState[];
      mission: MissionState;
      projectiles?: ProjectileState[]; // omitted when none in flight
      sentries?: SentryState[]; // deployed auto-turrets
      fires?: FireZoneState[]; // burning napalm zones
      nests?: NestState[]; // live destructible nests
      boss?: { id: number; hp: number; hpMax: number } | null; // titan healthbar
    }
  | { type: 'fired'; id: string; origin: Vec3; dir: Vec3; hit: Vec3 | null; hitKind: HitKind }
  | { type: 'sentryFire'; id: number; from: Vec3; to: Vec3 } // turret tracer
  | { type: 'recon'; pos: Vec3 } // recon pulse ping (clients reveal nearby bugs)
  | { type: 'bugDeath'; id: number; pos: Vec3; kind: number; by?: string } // by = killer player id
  | { type: 'splat'; pos: Vec3; kind: number } // acid projectile impact
  | { type: 'nestDeath'; i: number; pos: Vec3 } // a nest was sealed/destroyed
  | { type: 'boss'; pos: Vec3 } // titan arrival
  | { type: 'hit'; id: string; hp: number } // a player took damage
  | { type: 'down'; id: string }
  | { type: 'strat'; id: string; kind: string; target: Vec3; at: number } // impact at server time
  | { type: 'boom'; kind: string; pos: Vec3 } // ORBITAL | RESUPPLY | REINFORCE | SHUTTLE
  | { type: 'reinforced'; ids: string[]; pos: Vec3 }
  | { type: 'phase'; mission: MissionState }
  | { type: 'boarded'; id: string }
  // post-mission rewards: XP gained (with a breakdown), the updated profile, and
  // any weapons newly unlocked by a level-up this mission.
  | {
      type: 'progress';
      xpGained: number;
      leveledTo?: number; // present only when the player gained a level
      unlockedNew: string[]; // weapon ids newly available
      breakdown: { label: string; xp: number }[];
      profile: ProfileState;
    }
  | { type: 'pong'; t: number; now: number };

export function parseClientMsg(raw: string): ClientMsg | null {
  try {
    const msg = JSON.parse(raw);
    if (typeof msg !== 'object' || msg === null || typeof msg.type !== 'string') return null;
    return msg as ClientMsg;
  } catch {
    return null;
  }
}
